Exit the process gracefully

chaoyli opened the pull request:
It's to used to detect memory leak upon exiting the process.
Every thread in StorageEngine is already check the existed state.
If the existed state is set to be true, the thread will be exited.
But the the thread and the class in StarRocks is not seperated very well.
So the pull request do the best effort to the graceful exiting.

chaoyli committed:
Exit the process gracefully.
It's to used to detect memory leak upon exiting the process.
Every thread in StorageEngine is already check the existed state.
If the existed state is set to be true, the thread will be exited.
But the the thread and the class in StarRocks is not seperated very well.
So the pull request do the best effort to the graceful exiting.

imay commented on be/src/runtime/exec_env.h:
```diff
@@ -74,7 +74,7 @@ class ExecEnv {
     // Initial exec enviorment. must call this to init all
     static Status init(ExecEnv* env, const std::vector<StorePath>& store_paths);
-    static void destroy(ExecEnv* exec_env);
+    static void destory(ExecEnv* exec_env);
```
typo
```suggestion
    static void destroy(ExecEnv* exec_env);
```

chaoyli committed:
Exit the process gracefully.
It's to used to detect memory leak upon exiting the process.
Every thread in StorageEngine is already check the existed state.
If the existed state is set to be true, the thread will be exited.
But the the thread and the class in StarRocks is not seperated very well.
So the pull request do the best effort to the graceful exiting.

trueeyu commented on be/src/storage/merger.cpp:
```diff
@@ -87,6 +90,13 @@ OLAPStatus Merger::merge_rowsets(int64_t mem_limit, const TabletSharedPtr& table
+    }
+
+    if (bg_worker_stopped) {
+        std::cout << "after bg_worker_stopped is schema change:" << bg_worker_stopped << std::endl;
```
remove the log

chaoyli committed:
fix unit test bug

chaoyli committed:
Fix http server memory leak

chaoyli commented:
run starrocks_be_unittest

chaoyli committed:
Fix CurrentThread thread crash

chaoyli committed:
Fix unit test bug

sduzh commented on be/src/agent/agent_server.cpp:
```diff
@@ -96,7 +96,33 @@ AgentServer::AgentServer(ExecEnv* exec_env, const TMasterInfo& master_info)
 
-AgentServer::~AgentServer() = default;
+AgentServer::~AgentServer() {
+#ifndef STOP_POOL
```
This macro doesn't seem to make much sense

chaoyli committed:
Make thread been joinable

sduzh commented on be/src/agent/task_worker_pool.cpp:
```diff
@@ -169,6 +167,18 @@ void TaskWorkerPool::start() {
+    std::lock_guard l(_worker_thread_lock);
+    _worker_thread_condition_variable->notify_all();
+    for (uint32_t i = 0; i < _worker_count; ++i) {
+        _worker_threads[i].join();
```
I think you need to release the mutex first before joining the threads, otherwise, the threads cannot be waked up.

sduzh commented on be/src/runtime/current_thread.h:
```diff
@@ -19,6 +20,10 @@ class CurrentThread {
     ~CurrentThread() { commit(); }
 
     void commit() {
+        StorageEngine* storage_engine = ExecEnv::GetInstance()->storage_engine();
+        if (storage_engine != nullptr && storage_engine->bg_worker_stopped()) {
+            return;
+        }
```
What's the purpose of these tests here?

sduzh commented on be/src/storage/storage_engine.h:
```diff
@@ -283,7 +283,7 @@ class StorageEngine {
     std::unordered_map<std::string, RowsetSharedPtr> _unused_rowsets;
 
-    bool _stop_bg_worker = false;
+    bool _bg_worker_stopped = false;
```
Please use `std::atomic<bool>`, and use `std::memory_order_release` for writes and `std::memory_order_consume` for reads.

sduzh commented on be/src/storage/storage_engine.h:
```diff
@@ -171,14 +171,15 @@ class StorageEngine {
 
+    void stop();
+
+    bool bg_worker_stopped() { return _bg_worker_stopped; }
```
_bg_worker_stopped.load(std::memory_order_consume);

sduzh commented on be/src/storage/storage_engine.cpp:
```diff
@@ -454,24 +458,46 @@ bool StorageEngine::_delete_tablets_on_unused_root_path() {
+        _store_map.clear();
     }
-    _store_map.clear();
+    _bg_worker_stopped = true;
```
```suggestion
    _bg_worker_stopped.store(std::memory_order_release);
```

sduzh commented on be/src/storage/task/engine_clone_task.cpp:
```diff
@@ -646,6 +687,11 @@ Status EngineCloneTask::_clone_full_data(Tablet* tablet, TabletMeta* cloned_tabl
 }
 
 Status EngineCloneTask::_finish_clone_primary(Tablet* tablet, const std::string& clone_dir) {
+    bool bg_worker_stopped = ExecEnv::GetInstance()->storage_engine()->bg_worker_stopped();
+    if (!bg_worker_stopped) {
+        return Status::InternalError("Process is going to quit. The snapshot should be stopped as soon as possible.");
+    }
+
```
There is no while loop here, do we really need to check it here.

sduzh commented on be/src/storage/vectorized/compaction.cpp:
```diff
@@ -188,49 +189,66 @@ Status Compaction::merge_rowsets(int64_t mem_limit, Statistics* stats_output) {
-            } else {
-                return Status::InternalError("reader get_next error.");
+        bool bg_worker_stopped = ExecEnv::GetInstance()->storage_engine()->bg_worker_stopped();
+        while (!bg_worker_stopped) {
```
```suggestion
        while (!ExecEnv::GetInstance()->storage_engine()->bg_worker_stopped()) {
```
And remove the variable  `bg_worker_stopped`

chaoyli committed:
Make thread been joinable

chaoyli commented:
run starrocks_be_unittest

chaoyli committed:
Make thread been joinable

sduzh commented on be/src/runtime/external_scan_context_mgr.h:
```diff
@@ -51,10 +51,7 @@ class ExternalScanContextMgr {
-        _is_stop = true;
-        _keep_alive_reaper->join();
-    }
+    ~ExternalScanContextMgr() {}
```
Can you explain why you deleted the join?

sduzh commented on be/src/storage/olap_server.cpp:
```diff
@@ -188,7 +166,7 @@ void* StorageEngine::_base_compaction_thread_callback(void* arg, DataDir* data_d
     //TTabletId last_base_compaction_tablet_id = -1;
     Status status = Status::OK();
-    while (!_stop_bg_worker) {
+    while (!_bg_worker_stopped) {
```
Since you have using the `std::memory_order_release` to write `_bg_worker_stopped`, you should read it by `std::memory_order_consume` all the times. Maybe you can call `StorageEngine::bg_worker_stopped()` in these places.

chaoyli committed:
Make thread been joinable
